File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/qms/QmsBlackListFragment.java
package forpdateam.ru.forpda.ui.fragments.qms;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.AppCompatAutoCompleteTextView;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewStub;
import android.widget.ArrayAdapter;

import java.util.ArrayList;
import java.util.List;

import forpdateam.ru.forpda.App;
import forpdateam.ru.forpda.R;
import forpdateam.ru.forpda.api.others.user.ForumUser;
import forpdateam.ru.forpda.api.qms.interfaces.IQmsContact;
import forpdateam.ru.forpda.api.qms.models.QmsContact;
import forpdateam.ru.forpda.apirx.RxApi;
import forpdateam.ru.forpda.common.IntentHandler;
import forpdateam.ru.forpda.common.rx.Subscriber;
import forpdateam.ru.forpda.common.simple.SimpleTextWatcher;
import forpdateam.ru.forpda.ui.TabManager;
import forpdateam.ru.forpda.ui.fragments.RecyclerFragment;
import forpdateam.ru.forpda.ui.fragments.TabFragment;
import forpdateam.ru.forpda.ui.fragments.qms.adapters.QmsContactsAdapter;
import forpdateam.ru.forpda.ui.views.ContentController;
import forpdateam.ru.forpda.ui.views.DynamicDialogMenu;
import forpdateam.ru.forpda.ui.views.FunnyContent;

/**
 * Created by radiationx on 22.03.17.
 */

public class QmsBlackListFragment extends RecyclerFragment implements QmsContactsAdapter.OnItemClickListener<IQmsContact> {
    private AppCompatAutoCompleteTextView nickField;
    private QmsContactsAdapter adapter;
    private Subscriber<ArrayList<QmsContact>> mainSubscriber = new Subscriber<>(this);
    private DynamicDialogMenu<QmsBlackListFragment, IQmsContact> dialogMenu;
    private ArrayList<QmsContact> currentData;
    private Subscriber<List<ForumUser>> searchUserSubscriber = new Subscriber<>(this);

    public QmsBlackListFragment() {
        configuration.setDefaultTitle(App.get().getString(R.string.fragment_title_blacklist));
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        ViewStub viewStub = (ViewStub) findViewById(R.id.toolbar_content);
        viewStub.setLayoutResource(R.layout.toolbar_qms_black_list);
        viewStub.inflate();
        nickField = (AppCompatAutoCompleteTextView) findViewById(R.id.qms_black_list_nick_field);
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewsReady();
        nickField.addTextChangedListener(new SimpleTextWatcher() {
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                searchUser(s.toString());
            }
        });

        refreshLayout.setOnRefreshListener(this::loadData);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        adapter = new QmsContactsAdapter();
        recyclerView.setAdapter(adapter);
        adapter.setOnItemClickListener(this);
    }

    @Override
    protected void addBaseToolbarMenu() {
        super.addBaseToolbarMenu();
        getMenu().add(R.string.add)
                .setIcon(App.getVecDrawable(getContext(), R.drawable.ic_toolbar_add))
                .setOnMenuItemClickListener(item -> {
                    String nick = "";
                    if (nickField.getText() != null)
                        nick = nickField.getText().toString();
                    blockUser(nick);
                    return false;
                })
                .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS);
    }

    private void someClick(IQmsContact contact) {
        if (dialogMenu == null) {
            dialogMenu = new DynamicDialogMenu<>();
            dialogMenu.addItem(getString(R.string.profile), (context, data) -> IntentHandler.handle("https://4pda.ru/forum/index.php?showuser=" + data.getId()));
            dialogMenu.addItem(getString(R.string.dialogs), (context, data) -> {
                Bundle args = new Bundle();
                args.putString(TabFragment.ARG_TITLE, data.getNick());
                args.putInt(QmsThemesFragment.USER_ID_ARG, data.getId());
                args.putString(QmsThemesFragment.USER_AVATAR_ARG, data.getAvatar());
                TabManager.get().add(QmsThemesFragment.class, args);
            });
            dialogMenu.addItem(getString(R.string.delete), (context, data) -> context.unBlockUser(new int[]{data.getId()}));
        }
        dialogMenu.disallowAll();
        dialogMenu.allowAll();
        dialogMenu.show(getContext(), QmsBlackListFragment.this, contact);
    }


    @Override
    public boolean loadData() {
        if (!super.loadData()) {
            return false;
        }
        setRefreshing(true);
        mainSubscriber.subscribe(RxApi.Qms().getBlackList(), this::onLoadContacts, new ArrayList<>(), v -> loadData());
        return true;
    }

    private void onLoadContacts(ArrayList<QmsContact> data) {
        setRefreshing(false);
        if (data.isEmpty()) {
            if (!contentController.contains(ContentController.TAG_NO_DATA)) {
                FunnyContent funnyContent = new FunnyContent(getContext())
                        .setImage(R.drawable.ic_contacts)
                        .setTitle(R.string.funny_blacklist_nodata_title)
                        .setDesc(R.string.funny_blacklist_nodata_desc);
                contentController.addContent(funnyContent, ContentController.TAG_NO_DATA);
            }
            contentController.showContent(ContentController.TAG_NO_DATA);
        } else {
            contentController.hideContent(ContentController.TAG_NO_DATA);
        }
        recyclerView.scrollToPosition(0);
        currentData = data;
        adapter.addAll(currentData);
    }

    private void blockUser(String nick) {
        setRefreshing(true);
        mainSubscriber.subscribe(RxApi.Qms().blockUser(nick), this::onEditedList, currentData, null);
    }

    private void unBlockUser(int[] userIds) {
        setRefreshing(true);
        mainSubscriber.subscribe(RxApi.Qms().unBlockUsers(userIds), this::onEditedList, currentData, null);
    }

    private void onEditedList(ArrayList<QmsContact> data) {
        setRefreshing(false);
        if (currentData == data) return;
        currentData = data;
        adapter.addAll(currentData);
        nickField.setText("");
    }

    private void searchUser(String nick) {
        searchUserSubscriber.subscribe(RxApi.Qms().findUser(nick), this::onShowSearchRes, new ArrayList<>());
    }

    private void onShowSearchRes(List<ForumUser> res) {
        List<String> nicks = new ArrayList<>();
        for (ForumUser user : res) {
            nicks.add(user.getNick());
        }
        nickField.setAdapter(new ArrayAdapter<>(getContext(), android.R.layout.simple_dropdown_item_1line, nicks));
    }

    @Override
    public void onItemClick(IQmsContact item) {
        someClick(item);
    }

    @Override
    public boolean onItemLongClick(IQmsContact item) {
        someClick(item);
        return false;
    }
}

File: app/src/main/java/forpdateam/ru/forpda/ui/fragments/qms/QmsBlackListFragment.kt
package forpdateam.ru.forpda.ui.fragments.qms

import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.ViewStub
import android.widget.ArrayAdapter
import androidx.appcompat.widget.AppCompatAutoCompleteTextView
import androidx.recyclerview.widget.LinearLayoutManager
import forpdateam.ru.forpda.App
import forpdateam.ru.forpda.R
import forpdateam.ru.forpda.common.simple.SimpleTextWatcher
import forpdateam.ru.forpda.databinding.ToolbarQmsBlackListBinding
import forpdateam.ru.forpda.entity.remote.others.user.ForumUser
import forpdateam.ru.forpda.entity.remote.qms.QmsContact
import forpdateam.ru.forpda.presentation.qms.blacklist.QmsBlackListPresenter
import forpdateam.ru.forpda.presentation.qms.blacklist.QmsBlackListView
import forpdateam.ru.forpda.ui.fragments.RecyclerFragment
import forpdateam.ru.forpda.ui.fragments.qms.adapters.QmsContactsAdapter
import forpdateam.ru.forpda.ui.fragments.tabToolbarBinding
import forpdateam.ru.forpda.ui.views.ContentController
import forpdateam.ru.forpda.ui.views.DynamicDialogMenu
import forpdateam.ru.forpda.ui.views.FunnyContent
import forpdateam.ru.forpda.ui.views.adapters.BaseAdapter
import moxy.presenter.InjectPresenter
import moxy.presenter.ProvidePresenter

/**
 * Created by radiationx on 22.03.17.
 */

class QmsBlackListFragment : RecyclerFragment(), BaseAdapter.OnItemClickListener<QmsContact>,
    QmsBlackListView {

    private val toolbarBinding by tabToolbarBinding(ToolbarQmsBlackListBinding::bind)

    private val nickField: AppCompatAutoCompleteTextView
        get() = toolbarBinding.qmsBlackListNickField

    private lateinit var adapter: QmsContactsAdapter
    private val dialogMenu = DynamicDialogMenu<QmsBlackListFragment, QmsContact>()

    @InjectPresenter
    lateinit var presenter: QmsBlackListPresenter

    @ProvidePresenter
    fun providePresenter(): QmsBlackListPresenter = QmsBlackListPresenter(
        App.get().Di().qmsInteractor,
        App.get().Di().router,
        App.get().Di().linkHandler,
        App.get().Di().errorHandler
    )

    init {
        configuration.defaultTitle = App.get().getString(R.string.fragment_title_blacklist)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        baseInflateToolbar(R.layout.toolbar_qms_black_list)
        setScrollFlagsEnterAlways()
        nickField.addTextChangedListener(object : SimpleTextWatcher() {
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                presenter.searchUser(s.toString())
            }
        })

        refreshLayout.setOnRefreshListener { presenter.loadContacts() }
        recyclerView.layoutManager = LinearLayoutManager(requireContext())

        dialogMenu.apply {
            addItem(getString(R.string.profile)) { _, data ->
                presenter.openProfile(data)
            }
            addItem(getString(R.string.dialogs)) { _, data ->
                presenter.openDialogs(data)
            }
            addItem(getString(R.string.delete)) { _, data ->
                presenter.unBlockUser(data.user.id)
            }
        }

        adapter = QmsContactsAdapter()
        recyclerView.adapter = adapter
        adapter.setOnItemClickListener(this)
    }

    override fun addBaseToolbarMenu(menu: Menu) {
        super.addBaseToolbarMenu(menu)
        menu.add(R.string.add)
            .setIcon(R.drawable.ic_toolbar_add)
            .setOnMenuItemClickListener {
                var nick = ""
                if (nickField.text != null)
                    nick = nickField.text.toString()
                presenter.blockUser(nick)
                false
            }
            .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS)
    }

    override fun showContacts(items: List<QmsContact>) {
        setRefreshing(false)
        if (items.isEmpty()) {
            if (!contentController.contains(ContentController.TAG_NO_DATA)) {
                val funnyContent = FunnyContent(requireContext())
                    .setImage(R.drawable.ic_contacts)
                    .setTitle(R.string.funny_blacklist_nodata_title)
                    .setDesc(R.string.funny_blacklist_nodata_desc)
                contentController.addContent(funnyContent, ContentController.TAG_NO_DATA)
            }
            contentController.showContent(ContentController.TAG_NO_DATA)
        } else {
            contentController.hideContent(ContentController.TAG_NO_DATA)
        }
        recyclerView.scrollToPosition(0)
        adapter.addAll(items)
    }

    override fun clearNickField() {
        nickField.setText("")
    }

    override fun showFoundUsers(items: List<ForumUser>) {
        val nicks = items.map { it.nick.orEmpty() }
        nickField.setAdapter(
            ArrayAdapter(
                requireContext(),
                android.R.layout.simple_dropdown_item_1line,
                nicks
            )
        )
    }

    override fun showItemDialogMenu(item: QmsContact) {
        dialogMenu.apply {
            disallowAll()
            allowAll()
            show(requireContext(), this@QmsBlackListFragment, item)
        }
    }

    override fun onItemClick(item: QmsContact) {
        presenter.onItemLongClick(item)
    }

    override fun onItemLongClick(item: QmsContact): Boolean {
        presenter.onItemLongClick(item)
        return false
    }
}
